### PR TITLE
Give dynamic shared buffers address expressions

### PR DESCRIPTION
**Pull request by jacobhinkle**
This is related to #221. Currently, dynamic shared memory is used in two ways: either implicitly by reductions or when we have a `TensorView` with `MemoryType::Shared`. Note that our IR does not explicitly represent reduction (or Welford) smem at all. I'll refer to the shared memory space reserved for reductions and welfords as "reduction smem" and to all the shared mem above it via `TensorView`s as `tv smem`.

Currently when we generate CUDA kernels, we look for reductions and Welford ops and find the largest data type and we reserve that amount times the block size at the beginning of the dynamic smem array `shared_mem`. We then set `smem_offset` to that size. When `tv smem` is defined in the kernel, we align the `smem_offset` address to 16 bytes and use it for the buffer definition, then we add the new buffer's size to `smem_offset` in the generated kernel.

This method makes memory re-use cumbersome. We currently support aliasing identically-sized buffers, but if we wanted to create a new allocation with a different size between previous memory locations, we could not do that since we don't represent the smem offsets anywhere.

This PR creates an `address()` attribute on `kir::Allocate` which holds a scalar expression for the number of bytes above `smem_offset` at which to start the allocation. When generating a kernel, we use these expressions to directly find the address of the new buffer, and we never modify `smem_offset`. During lowering, in the `reuseMemoryAllocations` pass, we assign those address `Val` attributes without re-using any `tv smem`.

Note that this PR could potentially increase register use since we don't do any explicit CSE on these expressions: if there are multiple of smem TVs their offset expressions can grow in size due to intermediate alignment expressions. However, I didn't observe any change in register usage in our  current tests.

### CODE REVIEW

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -437,12 +437,21 @@ std::vector<PolymorphicValue> BinaryOp::evaluate(
       return {ceildiv(lhs, rhs)};
       break;
     case BinaryOpType::And:
-      return {lhs && rhs};
+      if (lhs.is<bool>() && rhs.is<bool>()) {
+        return {lhs && rhs};
+      }
+      return {lhs & rhs};
```
Comment (jacobhinkle) [inline csrc/ir/nodes.cpp:443]: See #688

**Comment by jacobhinkle** [comment]
Before this PR, the kernel for the `FusionSmemAlignment_CUDA` test looked like
```c++
__global__ void kernel1(Tensor<float, 5, 5> T0, Tensor<float, 1, 1> T4) {
  alignas(16) extern __shared__ char array[];
  unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  Array<nvfuser_index_t, 5, 1> a0;
  a0 = (T0).alloc_stride;
  int64_t i1;
  i1 = a0[0];
  int64_t i2;
  i2 = a0[1];
  int64_t i3;
  i3 = a0[2];
  int64_t i4;
  i4 = a0[3];
  int64_t i5;
  i5 = a0[4];
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T5 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += (70 * sizeof(float));
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T6 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += (14 * sizeof(float));
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T7 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += (7 * sizeof(float));
  float T2[84];
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T8 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += (12 * sizeof(float));
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T9 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += (3 * sizeof(float));
 ```
After this PR this becomes
```c++
__global__ void kernel1(Tensor<float, 5, 5> T0, Tensor<float, 1, 1> T4) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO;
  Array<nvfuser_index_t, 5, 1> a0;
  a0 = (T0).alloc_stride;
  int64_t i1;
  i1 = a0[0];
  int64_t i2;
  i2 = a0[1];
  int64_t i3;
  i3 = a0[2];
  int64_t i4;
  i4 = a0[3];
  int64_t i5;
  i5 = a0[4];
  float* T5 = reinterpret_cast<float*>(array + smem_offset + 0);
  float* T6 = reinterpret_cast<float*>(array + smem_offset + (((0 + (70 * 4)) + 15) & (~15)));
  float* T7 = reinterpret_cast<float*>(array + smem_offset + ((((((0 + (70 * 4)) + 15) & (~15)) + (14 * 4)) + 15) & (~15)));
  float T2[84];
  float* T8 = reinterpret_cast<float*>(array + smem_offset + (((((((((0 + (70 * 4)) + 15) & (~15)) + (14 * 4)) + 15) & (~15)) + (7 * 4)) + 15) & (~15)));
  float* T9 = reinterpret_cast<float*>(array + smem_offset + ((((((((((((0 + (70 * 4)) + 15) & (~15)) + (14 * 4)) + 15) & (~15)) + (7 * 4)) + 15) & (~15)) + (12 * 4)) + 15) & (~15)));
```

Notice that the `smem_offset` variable is still used to reserve memory for reductions, but it is no longer incremented for tv smem allocations.

File: csrc/device_lower/pass/alias_memory.cpp [hunk 2]
```diff
@@ -1382,6 +1383,50 @@ class AllocationReuseModifier : private kir::ExprMutator {
...
+
+ private:
+  using kir::IrVisitor::handle;
+
+  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
+    if (alignment == 1) {
+      return addr;
+    }
+    auto n_minus_one = IrBuilder::create<Val>(alignment - 1);
+    return bitwise_and(add(addr, n_minus_one), bitwise_not(n_minus_one));
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/alias_memory.cpp:1403]: Could you use `IrBuilder::bitwiseAndExpr` instead of `bitwise_and`? Unfortunately our functions in `arith.h` does not support `DataType::Index`, and I have an ongoing PR that changes most indexing vals to `DataType::Index`. Same thing applies to all ops from `arith.h`.

**Comment by naoyam** [comment]
Is reusing the reduction shmem region part of the plan?

**Comment by jacobhinkle** [comment]
> Is reusing the reduction shmem region part of the plan?

Yes. My plan is currently to have a series of PRs:
- Give smem allocations address expressions (this PR)
- Create simple reuse algorithm for non-reduction smem using address expressions
- Make reduction smem buffers explicitly represented in the IR as TVs and ensure they are reused.

File: csrc/device_lower/pass/alias_memory.cpp [hunk 3]
```diff
@@ -1382,6 +1383,50 @@ class AllocationReuseModifier : private kir::ExprMutator {
...
+    current_offset = add(address, size_bytes);
+  }
+
+ private:
+  Val* current_offset = nullptr;
+};
+
 } // namespace
 
 std::vector<Expr*> reuseMemoryAllocations(const std::vector<Expr*>& exprs) {
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/alias_memory.cpp:1432]: Is it possible to move the `reuseMemoryAllocations` down after `commonScalarMap().initialize(exprs_unrolled_loops);` so that we can do index hoisting? (If the answer is yes, then let's do it in a separate PR, don't change anything here.)

File: csrc/device_lower/pass/alias_memory.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/device_lower/pass/alias_memory.cpp:1403]: Definitely should use `bitwiseAndExpr`. Thanks.

I think the bitwise ops in `arith.cpp` work with `DataType::Index` since they just check `isIntegralType` which should be true for `Index`. But I haven't tried.

File: csrc/device_lower/pass/alias_memory.cpp (hunk 3 above)
Comment (jacobhinkle) [reply csrc/device_lower/pass/alias_memory.cpp:1432]: Yeah I was wondering about this; since the expressions are just cumulative sums they have tons of common subexpressions. But wouldn't moving it down later mean we could not hoist these common expressions since they would not yet be defined?

Comment (zasdfgbnm) [reply csrc/device_lower/pass/alias_memory.cpp:1432]: > But wouldn't moving it down later mean we could not hoist these common expressions since they would not yet be defined?

I don't understand. What do you mean?

`commonScalarMap().initialize(exprs_unrolled_loops);` just initializes the data structure for index hoisting and figure out what scalars are already computed as part of fusion definition. Once this data structure is initialized, we can use `GpuLower::current()->commonScalarMap().hoistScalar(val, for_loops_);` and common subexpr elimination should be automatically handled.

Comment (zasdfgbnm) [reply csrc/device_lower/pass/alias_memory.cpp:1432]: We can change `alloc->setAddress(address);` into
```C++
alloc->setAddress(GpuLower::current()->commonScalarMap().hoistScalar(address, for_loop_));
```

**Comment by jacobhinkle** [comment]
Now that we use `SimplifyingIrBuilder` as well as `hoistScalars` which simplifies expressions, constant allocation addresses are now output as shallow constants. The example from above now contains
```c++
  float* T5 = reinterpret_cast<float*>(array + smem_offset + 0);                                                                 
  float* T6 = reinterpret_cast<float*>(array + smem_offset + 288);                                                               
  float* T7 = reinterpret_cast<float*>(array + smem_offset + 352);                                                               
  float T2[84];                                                                                                                  
  float* T8 = reinterpret_cast<float*>(array + smem_offset + 384);                                                               
  float* T9 = reinterpret_cast<float*>(array + smem_offset + 432);
```

On the other hand, for an example showing multiple dynamic allocations, we can look at `FusionSmemDynamicPersistentNorm_CUDA`. Before this PR, the kernel contains:
```c++
  smem_offset = alignBufferSize(smem_offset, 16);     
  float* T14 = reinterpret_cast<float*>(array + smem_offset);                                                                    
  smem_offset += (((ceilDiv(T0.logical_size[1], i4)) * i4) * sizeof(float));
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T4 = reinterpret_cast<float*>(array + smem_offset);    
  smem_offset += (((ceilDiv(T0.logical_size[1], i4)) * i4) * sizeof(float));
```
After this PR, this becomes
```c++
  float* T14 = reinterpret_cast<float*>(array + smem_offset + 0);
  float* T4 = reinterpret_cast<float*>(array + smem_offset + (((((ceilDiv(T0.logical_size[1], i4)) * i4) * 4) + 15) & -16));
```
On a bit of a tangent: it seems like this test is an example where we could be more aggressive with aliasing. We do not alias T4 with T14, but it should be safe here. That's a topic for another PR.

**Comment by jacobhinkle** [comment]
!build

File: csrc/device_lower/pass/alias_memory.cpp [hunk 4]
```diff
@@ -674,6 +675,22 @@ class AllocationInfoMap : private kir::IrVisitor {
     return alias_map_;
   }
 
+  //! Get an ordered vector of loops containing an allocation
+  std::vector<kir::ForLoop*> getLoopNest(kir::Allocate* alloc) const {
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/device_lower/pass/alias_memory.cpp:679]: Is there any reason for not using this?
https://github.com/NVIDIA/Fuser/blob/55495b5e60fca6f5736777c7f6cf798d44c7c6cd/csrc/kernel_ir_dispatch.h#L46

**Comment by jacobhinkle** [comment]
!build

File: csrc/device_lower/pass/alias_memory.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/device_lower/pass/alias_memory.cpp:679]: You're right there's no good reason not to use that. I was thinking we would potentially need this after the traversal, but we don't. I'll remove it which means we won't need to hold allocation_info_map_ either.

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/device_lower/pass/alias_memory.cpp [hunk 5]
```diff
@@ -1382,6 +1383,63 @@ class AllocationReuseModifier : private kir::ExprMutator {
...
+              << " of size " << size_bytes->toInlineString() << " for T"
+              << alloc->buffer()->name() << " which has dtype "
+              << alloc->buffer()->dtype() << std::endl;
+    }
+
+    current_address_ = SimplifyingIrBuilder::addExpr(address, size_bytes);
+  }
+
+ private:
+  Val* current_address_ = nullptr;
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/device_lower/pass/alias_memory.cpp:1440]: That is a good idea. I was thinking we will get rid of `smem_offset_` soon anyway, but we can still represent the smem base address with a `NamedScalar`.

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/device_lower/pass/alias_memory.cpp:1440]: Sure, no problem. Don't bother.
